manager/orchestrator/restart/restart.go

```diff
@@ -190,6 +197,85 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 	return nil
 }
 
+// TaskRestartDelay calculates and returns the next delay based on t's RestartPolicy.
+// randomize is true if a backoff policy is being used,  false if the RestartPolicy
+// is configured for a fixed restart delay
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) (time.Duration, bool, error) {
+	restartDelay := time.Duration(0)
+	backoff := &api.BackoffPolicy{
+		Factor: defaults.Service.Task.Restart.Backoff.Factor,
+		Max:    defaults.Service.Task.Restart.Backoff.Max,
+	}
+	// If no restart is defined we use the default exponential backoff configuration
+	if t.Spec.Restart != nil {
+		var delayErr error
+		if delay, delayErr := gogotypes.DurationFromProto(t.Spec.Restart.Delay); delayErr == nil {
+			restartDelay = delay
+		}
+		if t.Spec.Restart.Backoff == nil && t.Spec.Restart.Delay != nil {
+			// Use fixed restart delay instead of exponential
+			if delayErr != nil {
+				log.G(ctx).WithError(delayErr).Error("invalid restart delay; using default")
+				// We only use this default if the user has configured for fixed restart delay
+				restartDelay, delayErr = gogotypes.DurationFromProto(defaults.Service.Task.Restart.Delay)
+
+				// If user defined a delay without a backoff, but that delay and the
+				// default delay can't be read we fail fast instead of defaulting to
+				// a randomized exponential backoff.
+				if delayErr != nil {
+					return 0, false, errors.New("error parsing RestartPolicy")
+				}
+			}
+			return restartDelay, false, nil
+		}
+		// If the user defined backoff can be read we override default values
+		if _, err := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Factor); err == nil {
+			backoff.Factor = t.Spec.Restart.Backoff.Factor
+		}
+		if _, err := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Max); err == nil {
+			backoff.Max = t.Spec.Restart.Backoff.Max
+		}
+	}
+
+	// Use jittered binary exponential backoff delay
+	var factor, max time.Duration
+	var err error
+	if factor, err = gogotypes.DurationFromProto(backoff.Factor); err != nil {
+		return 0, false, errors.New("error parsing backoff Factor")
+	}
+	if max, err = gogotypes.DurationFromProto(backoff.Max); err != nil {
+		return 0, false, errors.New("error parsing backoff Max")
+	}
+
+	// Get the failure count from history
+	var failures uint64
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	serviceID := t.ServiceID
+	tuple := orchestrator.SlotTuple{
+		Slot:      t.Slot,
+		ServiceID: t.ServiceID,
+		NodeID:    t.NodeID,
+	}
+	if r.historyByService[serviceID] != nil &&
+		r.historyByService[serviceID][tuple] != nil {
+		restartInfo := r.historyByService[serviceID][tuple]
+		failures = restartInfo.failuresSinceSuccess
+	} else {
+		// History is not set on the first run of a task
+		failures = 0
+	}
+
+	// Use floating point values to detect overflow.  math.Pow returns +Inf on overflow
+	backoffDuration := restartDelay + factor*time.Duration(math.Pow(2, float64(failures)))
```

Comment:
I think the best way to avoid overflow is to have an explicit limit on the exponent. Integer math should work fine.